### ISSUE TITLE
Importing cartopy.crs results in a DLL load fail on Python 3.8

### ISSUE THREAD

**Issue by geoffrey464** [SciTools/cartopy#1678]
### Description
Hello! I ran into an import issue today that I believe was caused by the new DLL loading scheme in Python 3.8.

A quick summary for those unfamiliar: Python 3.8 [Issue 36085](https://bugs.python.org/issue36085) introduced a security fix by disabling use of the Windows PATH environment variables to determine the paths for relevant DLLs. This introduced the function `os.add_dll_directory()` as a solution to add directories that were not being searched for DLLs due to the above security fix.

The above is paraphrased from the Issue messages,

> So the fundamental problem is that the default DLL search path on Windows changes in various contexts, and the only consistent approach is the most difficult to support with current packaging tools. The result is .pyd files that need to resolve .dll dependencies from directories *other* than where the .pyd file is located.

From the 3.8 [changelog](https://docs.python.org/3/whatsnew/3.8.html#bpo-36085-whatsnew)

> DLL dependencies for extension modules and DLLs loaded with ctypes on Windows are now resolved more securely. Only the system paths, the directory containing the DLL or PYD file, and directories added with add_dll_directory() are searched for load-time dependencies. Specifically, PATH and the current working directory are no longer used, and modifications to these will no longer have any effect on normal DLL resolution. If your application relies on these mechanisms, you should check for add_dll_directory() and if it exists, use it to add your DLLs directory while loading your library. Note that Windows 7 users will need to ensure that Windows Update KB2533623 has been installed (this is also verified by the installer). (Contributed by Steve Dower in bpo-36085.)

https://docs.python.org/3/library/os.html#os.add_dll_directory

This update appears to have broken at least the crs import in the Cartopy module.

I was able to get around this issue by including `os.add_dll_directory("<Path To GDAL DLLs>")` before the import statement below, but that seems to be a poor hack. It would be much cleaner if on build of the _crs*.pyd module it automatically imported the relevant DLLs. I would love to hear your input on this issue or if I'm going about it completely the wrong way on the installation. I'd prefer to not use a package manager such as conda if I can help it.

#### Code to reproduce

`import cartopy.crs as ccrs`

#### Traceback 

```
Traceback (most recent call last):
  File "cartopy_plot_data.py", line 6, in <module>
    import cartopy.crs as ccrs
  File "C:\Program Files\Python38\lib\site-packages\cartopy\__init__.py", line 96, in <module>
    import cartopy.crs
  File "C:\Program Files\Python38\lib\site-packages\cartopy\crs.py", line 36, in <module>
    from cartopy._crs import (CRS, Geodetic, Globe, PROJ4_VERSION,
ImportError: DLL load failed while importing _crs: The specified module could not be found.
```

#### Other possibly related issues
#1619 
#1562 

<details>
 <summary> Full environment definition</summary>

I have GDAL installed using the OSGeo4W64 package. Version GDAL 3.1.4, released 2020/10/20
The DLLs, libraries, and includes are in C:/OSGeo4W64/bin, C:/OSGeo4W64/lib, C:/OSGeo4W64/include respectively.
I installed cartopy as follows,
```
pip install --global-option=build_ext --global-option="-LC:/OSGeo4W64/lib" --global-option="-IC:/OSGeo4W64/include" cartopy==0.18.0
```

### Operating system & Python Version

Windows 10 64 bit, Python 3.8.6

### Pip Things

Cartopy==0.18.0
GDAL==3.1.4
pyshp==2.1.2
numpy==1.19.3
Cython==0.29.21
Shapely==1.7.1
six==1.15.0
matplotlib==3.3.2
Pillow==8.0.1
scipy==1.5.4
pykdtree==1.3.3

### Non-PipThings

GEOS==3.8.1-1
Proj==6.3.2-1
Proj-Dev==8.0.0-6

</details>

**Comment by dopplershift** [comment]
How did you install GDAL? How did you install cartopy and link against that library?
Can you import/use shapely with no problem?

**Comment by geoffrey464** [comment]
Hi @dopplershift I installed GDAL using the OSGeo4W64 package manager. 

This (defaulty) installs the GDAL executables, dlls, bats to C:\OSGeo4W64\bin, the c++ header files to C:\OSGeo4W64\include, and the library files to C:\OSGeo4W64\lib.
I also have the GDAL_DATA=C:\OSGeo4W64\share\gdal, GDAL_DRIVER_PATH=C:\OSGeo4W64\bin\gdalplugins, and USE_PATH_FOR_GDAL_PYTHON=YES environment variables set.

I then installed cartopy with the above GDAL library using, `pip install --global-option=build_ext --global-option="-LC:/OSGeo4W64/lib" --global-option="-IC:/OSGeo4W64/include" cartopy==0.18.0`

I can run the full shapely test suite with no problem.

**Comment by dopplershift** [comment]
Why did you need to pass those options to `pip install`? Cartopy doesn't link directly to gdal--in fact, I'm confused that doing anything for a DLL path and GDAL is necessary to import the `crs` submodule.

Can you try reinstalling cartopy with just `pip remove cartopy` and `pip install cartopy==0.18.0`?

**Comment by geoffrey464** [comment]
When I try to install cartopy without linking to the GEOS libraries/headers (which happen to be in the same location as the GDAL libraries), then I get build errors,
`lib/cartopy/trace.cpp(665): fatal error C1083: Cannot open include file: 'geos_c.h': No such file or directory`
Is there a better way to link the GEOS libraries on install?

**Comment by dopplershift** [comment]
Ah, that's where GEOS is...that's makes sense why it's necessary. I'm really not sure what CartoPy should do for this, though. It seems like installing the OSGeo stuff should make sure those libraries end up in a standard search path. I'm also not sure why this isn't an issue for us on 3.8 with conda.

**Comment by tkoch96** [comment]
Same issue here with Python 3.9.2. Is there any work around? Any way to use cartopy on windows?

**Comment by dopplershift** [comment]
@tkoch96 The easiest way is to install Anaconda/Miniconda, which makes it easy to get dependencies like GEOS and PROJ built and installed. 

I'm not an expert on dynamic library loading on Windows, so I'm not sure what the core issue is regarding the OSGeo install and CartoPy.

**Comment by ghuidobr** [comment]
Hi, I'm having the same problem, trying to import Cartopy and getting "DLL load failed while importing _crs: The specified module could not be found". I actually had Cartopy installed and working perfectly until yesterday but I had to uninstall and install Anaconda again and now it's not working. This is with Anaconda just installed, a conda env just created and just trying to install Cartopy on it using either conda or conda-forge.
I know very little about these stuff and I've tried to follow some of the above solutions but I can't find how to fix it. (How could I find the "Path To GDAL DLLs" for example?)

**Comment by greglucas** [comment]
@ghuidobr, did you activate your conda virtual environment that you installed Cartopy into? `conda activate cartopy-venv`, or if you're using an IDE there is usually some dropdown to switch environments. That should take care of all the paths and libraries for you I think.

**Comment by ghuidobr** [comment]
@greglucas Thanks for the response. I'm using jupyter lab, and I followed some steps so that the kernels shows up correctly. I'm able to open my notebook, select the kernel for the environment where Cartopy is installed, and I can import other packages like numpy or pyplot without problem but Cartopy gives me that error. But looking for solutions I realized that I might have a problem with malware, since my anaconda prompt doesn't work and neither does the windows cmd (they flash and close, and even the anaconda powershell has some strange behaviors like shrinking and erasing the prompt screen)(maybe the malware also messed with some paths?). I'll try to fix than and then see if my other problems go away.

**Comment by ChouhanDheeraj** [comment]
Hello,  I am working with python 3.9.4, I installed all the required library but still I got error.
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[30], line 1
----> 1 import cartopy

File c:\users\dheeraj chouhan\appdata\local\programs\python\python39\lib\site-packages\cartopy\__init__.py:110
    105     pass
    108 # Commonly used sub-modules. Imported here to provide end-user
    109 # convenience.
--> 110 import cartopy.crs
    111 import cartopy.feature  # noqa: F401  (flake8 = unused import)

File c:\users\dheeraj chouhan\appdata\local\programs\python\python39\lib\site-packages\cartopy\crs.py:27
     24 from pyproj.exceptions import ProjError
     25 from shapely.prepared import prep
---> 27 import cartopy.trace
     30 try:
     31     # https://github.com/pyproj4/pyproj/pull/912
     32     from pyproj.crs import CustomConstructorCRS as _CRS

ImportError: DLL load failed while importing trace: The specified module could not be found.

I don't know how to resolve this issue. I am trying to resolve this from last 1 week still I didn't get it. It stuck my project.
Can somebody help ? Please !

**Comment by ChouhanDheeraj** [comment]
> @greglucas Thanks for the response. I'm using jupyter lab, and I followed some steps so that the kernels shows up correctly. I'm able to open my notebook, select the kernel for the environment where Cartopy is installed, and I can import other packages like numpy or pyplot without problem but Cartopy gives me that error. But looking for solutions I realized that I might have a problem with malware, since my anaconda prompt doesn't work and neither does the windows cmd (they flash and close, and even the anaconda powershell has some strange behaviors like shrinking and erasing the prompt screen)(maybe the malware also messed with some paths?). I'll try to fix than and then see if my other problems go away.


Have u resolve this eroor ?

**Comment by ghuidobr** [comment]
@ChouhanDheeraj I'm sorry, this was a couple of years ago and I don't remember how I solved it. It was probably an issue with malware, as I mentioned in my last comment. But I haven't had the issue again since then. I'm sorry I can't be of more help.

**Comment by ChouhanDheeraj** [comment]
> @ChouhanDheeraj I'm sorry, this was a couple of years ago and I don't remember how I solved it. It was probably an issue with malware, as I mentioned in my last comment. But I haven't had the issue again since then. I'm sorry I can't be of more help.

Can u tell me what type of malware is it because of your window fireball, which blocking files or something else. Do u remember if possible and also can u suggest something else that can work at the position of cartopy. I have .nc file and I have to plot lwe_thickness data. I know that not a right but can u please help. I am beginner  in this. Please !!

**Comment by lgolston** [comment]
@ChouhanDheeraj - please provide more information on how you are installing cartopy. Is this with conda? What is the output of conda list?

**Comment by ChouhanDheeraj** [comment]
> @ChouhanDheeraj - please provide more information on how you are installing cartopy. Is this with conda? What is the output of conda list?

I installed cartopy with python pip install cartopy 
I also installed required libraries from pip

**Comment by ChouhanDheeraj** [comment]
> ![cartopt](https://user-images.githubusercontent.com/97750170/244877983-7fcfd9b6-654e-45f1-8a87-0b4ae24a0e76.jpg) ![cartopt](https://user-images.githubusercontent.com/97750170/244877983-7fcfd9b6-654e-45f1-8a87-0b4ae24a0e76.jpg)
> 
> > @ChouhanDheeraj - please provide more information on how you are installing cartopy. Is this with conda? What is the output of conda list?
> 
> I installed cartopy with python pip install cartopy I also installed required libraries from pip
> 
> I attached ss of files downloaded

@lgolston @greglucas

**Comment by ChouhanDheeraj** [comment]
I found there is error importing cartopy.trace @lgolston

**Comment by lgolston** [comment]
geos may be the problem; it needs to be version >=3.7.2. There is some discussion here: https://stackoverflow.com/questions/70177062/cartopy-not-able-to-identify-geos-for-proj-install-on-windows

**Comment by greglucas** [comment]
We have just released v0.22 which should help with the compatibility between packages and installation much easier. Please open a new issue if you are still having problems.
